src/others/contexts/file-downloader.tsx

```diff
@@ -0,0 +1,138 @@
+import { createContext, useState, useContext, FunctionComponent, useEffect } from "react";
+import { decodeCategory, decodeLocation } from "../helpers/decode-aid-request";
+import { useAidRequestQuery, useLocationsQuery, useSuppliesQuery } from "./api";
+import { useFilter } from "./filter";
+
+type FormattedAidRequests = {
+  [city: string]: {
+    date: string;
+    requests: [
+      {
+        category: string;
+        amount: number;
+      }
+    ];
+  };
+};
+
+export interface FileDownloaderContextValue {
+  isButtonDisabled: boolean;
+  downloadAsJSON: () => void;
+  downloadAsCSV: () => void;
+}
+
+const initFileDownloaderContextValue: FileDownloaderContextValue = {
+  isButtonDisabled: true,
+  downloadAsCSV: () => {},
+  downloadAsJSON: () => {},
+};
+
+const FileDownloaderContext = createContext<FileDownloaderContextValue>(initFileDownloaderContextValue);
+
+const triggerFileDownload = (dataBlob: Blob, extension: string) => {
+  const url = URL.createObjectURL(dataBlob);
+  const link = document.createElement("a");
+  link.setAttribute("href", url);
+  link.setAttribute("download", `ukraine-requests.${extension}`);
+  link.click();
+
+  URL.revokeObjectURL(url);
+};
+
+export function useFileDownloader(): FileDownloaderContextValue {
+  return useContext(FileDownloaderContext);
+}
+
+export const FileDownloaderContextProvider: FunctionComponent = ({ children }) => {
+  const { data: cities, isSuccess: citiesSuccess } = useLocationsQuery();
+  const { data: supplies, isSuccess: suppliesSuccess } = useSuppliesQuery();
+  const { data: aidRequests, isSuccess: aidRequestsSuccess } = useAidRequestQuery();
+  const { getActiveFilterItems } = useFilter();
+
+  const [isButtonDisabled, setButtonDisabled] = useState<boolean>(true);
+
+  const activeCategoryFilters = getActiveFilterItems("Categories");
+  const activeDateFilters = getActiveFilterItems("Dates");
+
+  useEffect(() => {
+    setButtonDisabled(!citiesSuccess && !suppliesSuccess && !aidRequestsSuccess);
+  }, [citiesSuccess, suppliesSuccess, aidRequestsSuccess]);
+
+  const filteredAidRequests = aidRequests?.filter(
+    (aidRequest) =>
+      (activeCategoryFilters.length === 0 || activeCategoryFilters.includes(aidRequest.category_id)) &&
+      (activeDateFilters.length === 0 || activeDateFilters.includes(aidRequest.date))
+  );
+
+  const uniqueCategories: Set<string> = new Set();
+
+  const formattedAidRequests = filteredAidRequests?.reduce((result: FormattedAidRequests, request) => {
+    if (supplies && cities) {
+      const category = decodeCategory(supplies, request.category_id);
+      const city = decodeLocation(cities, request.city_id).name;
+
+      uniqueCategories.add(category);
+
+      const formattedRequest = {
+        category: category,
+        amount: request.requested_amount,
+      };
+
+      if (result[city]) {
+        result[city].requests.push(formattedRequest);
```

Comment:
i think there's a problem in this code with respect to "date"s.

in the filter i see that you assume that a multiple dates could be selected,
but here you are merging all requests into a dict value that has one specific date.

Reply:
let's have a nested dict? when inner one is keyed by date.
or the outer is by date and the inner by city?